Add support for RegExp lookbehind assertions

msaboff opened the pull request:
#### 46e6b3f97425a4a7bb16fc175288903a5f74d5f2
<pre>
Add support for RegExp lookbehind assertions
<a href="https://bugs.webkit.org/show_bug.cgi?id=174931">https://bugs.webkit.org/show_bug.cgi?id=174931</a>
rdar://33183185

This change implements RegExp lookbehind in the Yarr interpreter.

This change introduces the notion of match direction, either forward or backward.
The forward match direction is the way the current code works, matching disjunciton terms and the subject
string in a right to left manner.  Lookbehind assertions, as defined in the EcmaScript spec, process disjunctions
terms right to left matching the correspondding subject string right to left as well.

Except for the Yarr JIT, almost all of the Yarr code has been touched to account for this backward matching.
An additional Byteterm has been added, HaveCheckedInput, which checks that there is at least as many characters
available in the input stream, but it doesn&apos;t move the input stream position.  This is basically a CheckInput,
without moving the input position.  For variable counted terms, we still need to check that we won&apos;t try to access
characters beyond the first character of the subject string.  For functions like readSurrogatePairChecked(),
we check for input before calling the funcion.  For new input functions with a try prefix like tryReadBackward,
the function itselfs checks for available input.  After these checks prove that it is safe to access an offset
to the left of the current input position, the actual matching can be performed.

The Yarr parser, parses regular expression in left to right order.  It also computes character offest in forward
order.  When we Byteterm compile, we process backward matching disjunctions right to left.  The parser also has
special handling of forward references within a backward matching parenthetical group.  All such forward references
are saved for that parenthetical group and are processed at the end of the group.  Every one of these forward
reference are check to see if a capture to the right of the forward reference was found, if so the forward
reference is converted to a back reference.

As part of this work, the ByteTerm dumping code was significantly updated to allow for not only dumping of the
ByteCode after it has been generated, but to dump ByteCode while it is being interpreted.  This ByteTerm dumping
while interpreting is enabled with the Interpreter::verbose compile time constant.

Reviewed by Yusuke Suzuki.

* JSTests/stress/regexp-lookbehind.js: New tests.
(arrayToString):
(dumpValue):
(compareArray):
(testRegExp):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/RegExp.cpp:
(JSC::RegExp::compile):
(JSC::RegExp::compileMatchOnly):
* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::ByteTermDumper::ByteTermDumper):
(JSC::Yarr::ByteTermDumper::unicode):
(JSC::Yarr::Interpreter::InputStream::readForCharacterDump):
(JSC::Yarr::Interpreter::InputStream::tryReadBackward):
(JSC::Yarr::Interpreter::InputStream::tryUncheckInput):
(JSC::Yarr::Interpreter::InputStream::isValidNegativeInputOffset):
(JSC::Yarr::Interpreter::InputStream::dump const):
(JSC::Yarr::Interpreter::checkCharacter):
(JSC::Yarr::Interpreter::checkSurrogatePair):
(JSC::Yarr::Interpreter::checkCasedCharacter):
(JSC::Yarr::Interpreter::checkCharacterClass):
(JSC::Yarr::Interpreter::checkCharacterClassDontAdvanceInputForNonBMP):
(JSC::Yarr::Interpreter::tryConsumeBackReference):
(JSC::Yarr::Interpreter::matchAssertionWordBoundary):
(JSC::Yarr::Interpreter::backtrackPatternCharacter):
(JSC::Yarr::Interpreter::backtrackPatternCasedCharacter):
(JSC::Yarr::Interpreter::matchCharacterClass):
(JSC::Yarr::Interpreter::backtrackCharacterClass):
(JSC::Yarr::Interpreter::matchBackReference):
(JSC::Yarr::Interpreter::backtrackBackReference):
(JSC::Yarr::Interpreter::recordParenthesesMatch):
(JSC::Yarr::Interpreter::matchParenthesesOnceBegin):
(JSC::Yarr::Interpreter::matchParenthesesOnceEnd):
(JSC::Yarr::Interpreter::backtrackParenthesesOnceEnd):
(JSC::Yarr::Interpreter::matchParentheticalAssertionBegin):
(JSC::Yarr::Interpreter::backtrackParentheticalAssertionBegin):
(JSC::Yarr::Interpreter::matchDisjunction):
(JSC::Yarr::ByteCompiler::compile):
(JSC::Yarr::ByteCompiler::haveCheckedInput):
(JSC::Yarr::ByteCompiler::assertionWordBoundary):
(JSC::Yarr::ByteCompiler::atomPatternCharacter):
(JSC::Yarr::ByteCompiler::atomCharacterClass):
(JSC::Yarr::ByteCompiler::atomBackReference):
(JSC::Yarr::ByteCompiler::atomParenthesesOnceBegin):
(JSC::Yarr::ByteCompiler::atomParenthesesTerminalBegin):
(JSC::Yarr::ByteCompiler::atomParenthesesSubpatternBegin):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionBegin):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesSubpatternEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesOnceEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesTerminalEnd):
(JSC::Yarr::ByteCompiler::emitDisjunction):
(JSC::Yarr::ByteCompiler::isSafeToRecurse):
(JSC::Yarr::ByteTermDumper::dumpTerm):
(JSC::Yarr::ByteTermDumper::dumpDisjunction):
(JSC::Yarr::Interpreter::InputStream::readPair): Deleted.
(JSC::Yarr::ByteCompiler::dumpDisjunction): Deleted.
* Source/JavaScriptCore/yarr/YarrInterpreter.h:
(JSC::Yarr::ByteTerm::ByteTerm):
(JSC::Yarr::ByteTerm::HaveCheckedInput):
(JSC::Yarr::ByteTerm::WordBoundary):
(JSC::Yarr::ByteTerm::BackReference):
(JSC::Yarr::ByteTerm::isCharacterType):
(JSC::Yarr::ByteTerm::isCasedCharacterType):
(JSC::Yarr::ByteTerm::isCharacterClass):
(JSC::Yarr::ByteTerm::matchDirection):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::dumpCompileFailure):
* Source/JavaScriptCore/yarr/YarrJIT.h:
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::parseParenthesesBegin):
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::resetForReparsing):
(JSC::Yarr::YarrPatternConstructor::assertionBOL):
(JSC::Yarr::YarrPatternConstructor::atomPatternCharacter):
(JSC::Yarr::YarrPatternConstructor::atomBuiltInCharacterClass):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesSubpatternBegin):
(JSC::Yarr::YarrPatternConstructor::atomParentheticalAssertionBegin):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesEnd):
(JSC::Yarr::YarrPatternConstructor::atomBackReference):
(JSC::Yarr::YarrPatternConstructor::copyDisjunction):
(JSC::Yarr::YarrPatternConstructor::quantifyAtom):
(JSC::Yarr::YarrPatternConstructor::disjunction):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::SavedContext::SavedContext):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::SavedContext::restore):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::ParenthesisContext):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::push):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::pop):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::setInvert):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::invert const):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::setMatchDirection):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::matchDirection const):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::reset):
(JSC::Yarr::YarrPatternConstructor::pushParenthesisContext):
(JSC::Yarr::YarrPatternConstructor::popParenthesisContext):
(JSC::Yarr::YarrPatternConstructor::setParenthesisInvert):
(JSC::Yarr::YarrPatternConstructor::parenthesisInvert const):
(JSC::Yarr::YarrPatternConstructor::setParenthesisMatchDirection):
(JSC::Yarr::YarrPatternConstructor::parenthesisMatchDirection const):
(JSC::Yarr::YarrPattern::YarrPattern):
(JSC::Yarr::dumpCharacterClass):
(JSC::Yarr::PatternTerm::dump):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::PatternTerm::PatternTerm):
(JSC::Yarr::PatternTerm::convertToBackreference):
(JSC::Yarr::PatternTerm::setMatchDirection):
(JSC::Yarr::PatternTerm::matchDirection const):
(JSC::Yarr::PatternAlternative::PatternAlternative):
(JSC::Yarr::PatternAlternative::matchDirection const):
(JSC::Yarr::PatternDisjunction::addNewAlternative):
(JSC::Yarr::YarrPattern::resetForReparsing):
* Source/JavaScriptCore/yarr/YarrSyntaxChecker.cpp:
(JSC::Yarr::SyntaxChecker::atomParentheticalAssertionBegin):
* Source/WTF/wtf/PrintStream.cpp:
(WTF::printInternal):
* Source/WTF/wtf/PrintStream.h:
* Source/WebCore/contentextensions/URLFilterParser.cpp:
(WebCore::ContentExtensions::PatternParser::atomParentheticalAssertionBegin):

Canonical link: <a href="https://commits.webkit.org/257823@main">https://commits.webkit.org/257823@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/52844d018234f70d16597910045cd35d33352f5f

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100129 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9298 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33203 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109463 "Built successfully") | [❌ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169698 "Failed to checkout and rebase branch from PR 7109") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104124 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10181 "Built successfully") | [⏳ 🛠 mac-AS-debug ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92560 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107353 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105898 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7697 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90980 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34406 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89603 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/22375 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90710 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/3066 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/23890 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86674 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/497 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/3038 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2 ](https://ews-build.webkit.org/#/builders/macOS-AppleSilicon-Big-Sur-Debug-Build-EWS "Waiting in queue, processing has not started yet") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29050 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9168 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43375 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89555 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5374 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4876 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/20023 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -39,9 +39,37 @@
 template<typename CharType>
 class Interpreter {
 public:
+    static const bool verbose = false;
```
Use `static constexpr`.

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/runtime/RegExp.cpp:
```diff
@@ -262,7 +263,7 @@ void RegExp::compile(VM* vm, Yarr::CharSize charSize)
 
     if (Options::dumpCompiledRegExpPatterns())
-        dataLog("Can't JIT this regular expression: \"", m_patternString, "\"\n");
+        dataLog("Can't JIT this regular expression: \"/", m_patternString, "/\"\n");
```
Let's use `dataLogLnIf` for new change.

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/runtime/RegExp.cpp:
```diff
@@ -326,7 +328,7 @@ void RegExp::compileMatchOnly(VM* vm, Yarr::CharSize charSize)
 
     if (Options::dumpCompiledRegExpPatterns())
-        dataLog("Can't JIT this regular expression: \"", m_patternString, "\"\n");
+        dataLog("Can't JIT this regular expression: \"/", m_patternString, "/\"\n");
```
Let's use `dataLogLnIf` for new change.

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1736,7 +2058,7 @@ class ByteCompiler {
 
 #ifndef NDEBUG
```
Use ASSERT_ENABLED.

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -596,17 +600,22 @@ class YarrPatternConstructor {
     }
 
-    void atomParentheticalAssertionBegin(bool invert = false)
+    void atomParentheticalAssertionBegin(bool invert = false, MatchDirection matchDirection = Forward)
```
Because this function is called only from 4 places, I rather would like to see that these parameters are always specified instead of default parameters to improve readability.

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -882,11 +934,14 @@ class YarrPatternConstructor {
-                error = setupDisjunctionOffsets(term.parentheses.disjunction, currentCallFrameSize + YarrStackSpaceForBackTrackInfoParentheticalAssertion, currentInputPosition, currentCallFrameSize);
-                if (hasError(error))
-                    return error;
+                {
```
Why not just place this brace just after `: ` to make this normal case clause?

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.h:
```diff
@@ -128,6 +128,12 @@ enum class QuantifierType : uint8_t {
     NonGreedy,
 };
 
+enum MatchDirection {
```
Let's make sure that this is `uint8_t` size with `: uint8_t`.

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.h:
```diff
@@ -325,11 +355,17 @@ struct PatternAlternative {
         return m_onceThrough;
     }
 
+    MatchDirection matchDirection()
```
const?

msaboff replied:
Fixed.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -770,18 +814,27 @@ class YarrPatternConstructor {
+                if (m_alternative->lastTerm().type == PatternTerm::Type::ParenthesesSubpattern)
+                    m_alternative->lastTerm().parentheses.isCopy = true;
+            } else {
+                PatternTerm& variableTerm = m_alternative->lastTerm();
```
I think this reference is wrong because we are adding a new `m_terms` before using this variableTerm. If Vector is expanded, this reference should point to the wrong memory.

msaboff replied:
Refactored to avoid modifying variableTerm after the append.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -637,19 +646,46 @@ class YarrPatternConstructor {
         }
 
+        m_pattern.m_containsBackreferences = true;
+
```
Don't we need to mark this before `if (subpatternId > m_pattern.m_numSubpatterns) {` check?

If we need this flag only when we actually emit backreference, so how about moving this to L710?

msaboff replied:
Moved.

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -637,19 +646,46 @@ class YarrPatternConstructor {
+
+        bool shouldTryConvertingForwardReferencesToBackreferences =
+            lastTerm.type == PatternTerm::Type::ParentheticalAssertion
+            && m_forwardReferencesInLookbehind.size()
```
Let's use `!isEmpty()` since it looks easier to read.

msaboff replied:
Changed

Constellation commented on Source/JavaScriptCore/yarr/YarrPattern.cpp:
```diff
@@ -662,6 +698,14 @@ class YarrPatternConstructor {
+                return;
+            }
         }
 
```
Should we put `m_pattern.m_containsBackreferences = true;` here instead?

msaboff replied:
I don't think so.  This could still be a forward reference that doesn't get converted.  The m_pattern.m_containsBackreferences = true around L663 takes care of the forward references that get converted.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.h:
```diff
@@ -317,11 +352,31 @@ struct ByteTerm {
 
+    bool isCharacterType()
+    {
+        return (type >= Type::PatternCharacterOnce && type <= Type::PatternCharacterNonGreedy);
```
Don't need parentheses.

msaboff replied:
Took care of this set and the next two.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.h:
```diff
@@ -317,11 +352,31 @@ struct ByteTerm {
+
+    bool isCasedCharacterType()
+    {
+        return (type >= Type::PatternCasedCharacterOnce && type <= Type::PatternCasedCharacterNonGreedy);
```
Don't need parentheses.

msaboff replied:
Agree.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.h:
```diff
@@ -317,11 +352,31 @@ struct ByteTerm {
+
+    bool isCharacterClass()
+    {
+        return (type == Type::CharacterClass);
```
Don't need parentheses.

msaboff replied:
Agreed

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1287,16 +1461,60 @@ class Interpreter {
-        if (!--remainingMatchCount)
+        if (!--remainingMatchCount) {
+            if (verbose)
+                dataLog("      Reached match limit - Returning ErrorHitLimit\n");
```
Let's use dataLogLnIf.

msaboff replied:
Changed.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1287,16 +1461,60 @@ class Interpreter {
+#define DUMP_EXTRA(...) \
+{ \
+    if (verbose) \
+        dataLog(" ", __VA_ARGS__); \
```
Let's use dataLogIf.

msaboff replied:
Changed.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1287,16 +1461,60 @@ class Interpreter {
+#define DUMP_EXTRA_IF(predicate, ...) \
+{ \
+    if (verbose && predicate) \
+        dataLog(" ", __VA_ARGS__); \
```
Use dataLogIf

msaboff replied:
Changed.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1662,6 +1981,9 @@ class Interpreter {
         DisjunctionContext* context = allocDisjunctionContext(pattern->m_body.get());
 
+        if (verbose)
+            dataLog("  Interpret input: ", input, "\n  Matching\n");
```
dataLogLnIf.

msaboff replied:
Changed.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -299,6 +363,24 @@ class Interpreter {
             return (((pos + offset) <= length) && ((pos + offset) >= pos));
         }
 
+        bool isValidOffset(unsigned offset)
```
We should name it `negativeInputOffset` and `isValidNegativeInputOffset`.

msaboff replied:
Renamed `neagativeInputOffset` to `isValidNegativeInputOffset`.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -408,47 +490,76 @@ class Interpreter {
+        if (!input.isValidOffset(negativeInputOffset))
+            return false;
+
+        return term.atom.patternCharacter == input.tryReadBackward(negativeInputOffset);
```
Why don't we have `readBackwardChecked` and use it instead?
I see that readSurrogatePairChecked and readBackwardSurrogatePairChecked are paired. But it seems that readChecked's pair, `readBackwardChecked` doesn't exist.

msaboff replied:
Given that any backward matching term could be preceded by a term with a variable count, we need to check before reading a character that we have input when we read backward.  Thus the name `tryReadBackward`.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -408,47 +490,76 @@ class Interpreter {
-        return (loChar == ch) || (hiChar == ch);
+        ASSERT(term.isCasedCharacterType());
+
+        int ch = term.matchDirection() == Forward ? input.readChecked(negativeInputOffset) : input.tryReadBackward(negativeInputOffset);
```
Ditto, why isn't it `readBackwardChecked`?

msaboff replied:
See above.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -408,47 +490,76 @@ class Interpreter {
-        int inputChar = input.readChecked(negativeInputOffset);
+        ASSERT(term.isCharacterClass());
+
+        int inputChar = term.matchDirection() == Forward ? input.readChecked(negativeInputOffset) : input.tryReadBackward(negativeInputOffset);
```
Ditto.

msaboff replied:
DItto.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -408,47 +490,76 @@ class Interpreter {
+        if (term.matchDirection() == Forward)
+            readCharacter = characterClass->hasOnlyNonBMPCharacters() ? input.readSurrogatePairChecked(negativeInputOffset) :  input.readChecked(negativeInputOffset);
+        else
+            readCharacter = characterClass->hasOnlyNonBMPCharacters() ? input.readBackwardSurrogatePairChecked(negativeInputOffset) :  input.tryReadBackward(negativeInputOffset);
```
Ditto. And if we need to additional check for backwards, then why is `readBackwardSurrogatePairChecked` paired with `readSurrogatePairChecked`? Shouldn't it be `tryReadBackwardSurrogatePair`?

msaboff replied:
`readBackwardSurrogatePairChecked()` is no longer used in my working code.  I deleted it.

Constellation replied:
OK, so I will see after the patch is updated.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -457,7 +568,7 @@ class Interpreter {
                 ++i;
             } else
-                ch = input.readChecked(negativeInputOffset + matchSize - i);
+                ch = term.matchDirection() == Forward ? input.readChecked(negativeInputOffset + matchSize - i) : input.tryReadBackward(negativeInputOffset + matchSize - i);
```
Ditto.

msaboff replied:
It was deleted.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1347,43 +1570,75 @@ class Interpreter {
+                    position = input.getPos();
+                }
+            } else {
+                auto term = currentTerm();
```
Use `const auto&`

msaboff replied:
That change would require refactoring const into all the match* and backtrack* functions as well as the accessors for ByteTerm.

We could make that change later.

Constellation replied:
OK, so let's make it `auto&` for now.

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1392,46 +1647,85 @@ class Interpreter {
+                        }
+                    }
+                } else {
+                    auto term = currentTerm();
```
Ditto.

msaboff replied:
See above.

Constellation replied:
Let's make it `auto&`.

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1392,46 +1647,85 @@ class Interpreter {
+                
+                MATCH_NEXT();
+            } else {
+                auto term = currentTerm();
```
Ditto.

msaboff replied:
See above.

Constellation replied:
Let's make it `auto&`

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1287,16 +1461,60 @@ class Interpreter {
 
-#define MATCH_NEXT() { ++context->term; goto matchAgain; }
-#define BACKTRACK() { --context->term; goto backtrack; }
+#define MATCH_NEXT() { if (verbose) { dataLog(" - Match\n"); if (btrack) { dataLog("  Matching\n"); btrack = false; } }  ++context->term; goto matchAgain; }
+#define BACKTRACK() { if (verbose) { dataLog(" - Backtrack\n"); if (!btrack) { dataLog("  Backtracking\n"); btrack = true; } } --context->term; goto backtrack; }
```
Use `dataLogLnIf` for verbose case.

msaboff replied:
The `verbose` case extends to the next if in both macros.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1287,16 +1494,59 @@ class Interpreter {
+        if (ch == -1) \
+            dataLog("<illegal>"); \
+        else if (ch < 0x1000)\
+            dataLog(static_cast<wchar_t>(ch)); \
```
Let's use `char16_t` instead.

msaboff replied:
We don't have a printStream method for `char16_t`, only `wchar_t`.  I did add another `0` to the comparison value on the line above so that we can print all BMP characters.

msaboff replied:
Changed the `printInternal()` method and this use to be `char16_t`.

Constellation commented on Source/WTF/wtf/PrintStream.h:
```diff
@@ -119,6 +119,7 @@ inline void printInternal(PrintStream& out, UniquedStringImpl& value) { printInt
 WTF_EXPORT_PRIVATE void printInternal(PrintStream&, bool);
 WTF_EXPORT_PRIVATE void printInternal(PrintStream&, signed char); // NOTE: this prints as a number, not as a character; use CharacterDump if you want the character
 WTF_EXPORT_PRIVATE void printInternal(PrintStream&, unsigned char); // NOTE: see above.
+WTF_EXPORT_PRIVATE void printInternal(PrintStream&, wchar_t);
```
Let's define `char16_t` thing.

msaboff replied:
Done

Constellation commented on Source/WTF/wtf/PrintStream.cpp:
```diff
@@ -144,6 +144,11 @@ void printInternal(PrintStream& out, unsigned char value)
     out.printf("%u", static_cast<unsigned>(value));
 }
 
+void printInternal(PrintStream& out, wchar_t value)
```
Let's use `char16_t`.

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -408,56 +471,80 @@ class Interpreter {
-                ch = input.readSurrogatePairChecked(negativeInputOffset + matchSize - i);
+                if (term.matchDirection() == Backward && negativeInputOffset > input.getPos())
+                    return false;
+
```
Why do we need to have this check again while L535 is doing the same check, and reread does not change input position?

msaboff replied:
I don't think we need it.  Removed.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -517,19 +611,38 @@ class Interpreter {
-                input.uncheckInput(U16_LENGTH(term.atom.patternCharacter));
+                if (term.matchDirection() == Forward)
+                    input.uncheckInput(U16_LENGTH(term.atom.patternCharacter));
+                else if (!input.checkInput(U16_LENGTH(term.atom.patternCharacter)))
```
We should rename `checkInput` to `tryCheckInput` since this is doing a check as the same way to `tryUncheckInput`.

Constellation replied:
Also, let's change the organization of code to make Forward/Backward case clear.

```
if (term.matchDirection() == Forward)
    input.uncheckInput(U16_LENGTH(term.atom.patternCharacter));
else {
    if (!input.tryCheckInput(U16_LENGTH(term.atom.patternCharacter)))
        break;
}
```

msaboff replied:
I restructured the `if else` as suggested.

checkInput is a preexisting construct.  There is a Byteterm by the same name.  It always had a **can fail** semantic.  That refactoring would confuse my historic understanding of the code.  That refactoring should also be done in  YarrJIT.cpp as well.  When we do the YarrJIT work, that may be a good to to refactor all the code.

Constellation replied:
I think it is very confusing that `checkInput` can fail, `uncheckInput` cannot fail, `tryUncheckInput` can fail.
So, if you think we should not change `checkInput` name, then we should rename `tryUncheckInput` to `uncheckInput`, and rename `uncheckInput` to `uncheckInputWithoutFailure` etc.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -546,15 +659,28 @@ class Interpreter {
+                if (term.matchDirection() == Forward)
+                    input.uncheckInput(1);
+                else if (!input.checkInput(1))
+                    break;
```
Ditto, let's a bit reorganize this code to make Forward/Backward clear.

```
if (term.matchDirection() == Forward)
    input.uncheckInput(1);
else {
    if (!input.tryCheckInput(1))
        break;
}
```

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -1347,43 +1602,90 @@ class Interpreter {
+                    }
+                }
+            } else {
+                auto term = currentTerm();
```
Use `auto&`

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -2077,40 +2466,60 @@ class ByteCompiler {
+            auto termCount = alternative->m_terms.size();
+            for (unsigned i = 0; i < termCount; ++i) {
+                unsigned termIndex = matchDirection == Forward ? i : termCount - 1 - i;
+                auto term = alternative->m_terms[termIndex];
```
Let's make it `auto&`.

msaboff replied:
Done.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -495,10 +587,12 @@ class Interpreter {
 
```
For `matchAssertionBOL`, etc. don't we need to use tryReadBackward to check `term.inputPosition` is valid?

msaboff replied:
The last major refactoring I did was to align the offsets for forward and backward matching.  This allowed some of the code to not need changes like `matchAssertionBOL` here.  I believe we don't need the check.

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -517,19 +611,38 @@ class Interpreter {
+                if (term.matchDirection() == Forward)
+                    input.uncheckInput(U16_LENGTH(term.atom.patternCharacter));
+                else if (!input.checkInput(U16_LENGTH(term.atom.patternCharacter)))
+                    break;
```
Let's reorganize this code to make Forward/Backward clear.
```
if (term.matchDirection() == Forward)
    input.uncheckInput(U16_LENGTH(term.atom.patternCharacter));
else {
    if (!input.tryCheckInput(U16_LENGTH(term.atom.patternCharacter)))
        break;
}
```

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -546,15 +659,28 @@ class Interpreter {
+                if (term.matchDirection() == Forward)
+                    input.uncheckInput(1);
+                else if (!input.checkInput(1))
+                    break;
```
Ditto.

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -604,16 +762,30 @@ class Interpreter {
-                position = input.getPos();
-            }
-            backTrack->matchAmount = matchAmount;
+                backTrack->matchAmount = matchAmount;
```
`return true;` here to make this code aligned to the other code.

msaboff replied:
Done

Constellation commented on Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
```diff
@@ -604,16 +762,30 @@ class Interpreter {
+                    position = input.getPos();
+                }
+                backTrack->matchAmount = matchAmount;
+            }
```
And remove this one-level nest.

msaboff replied:
.. and done.

Constellation reviewed (approved):
r=me

msaboff committed:
Add support for RegExp lookbehind assertions
https://bugs.webkit.org/show_bug.cgi?id=174931
rdar://33183185

This change implements RegExp lookbehind in the Yarr interpreter.

This change introduces the notion of match direction, either forward or backward.
The forward match direction is the way the current code works, matching disjunciton terms and the subject
string in a right to left manner.  Lookbehind assertions, as defined in the EcmaScript spec, process disjunctions
terms right to left matching the correspondding subject string right to left as well.

Except for the Yarr JIT, almost all of the Yarr code has been touched to account for this backward matching.
An additional Byteterm has been added, HaveCheckedInput, which checks that there is at least as many characters
available in the input stream, but it doesn't move the input stream position.  This is basically a CheckInput,
without moving the input position.  For variable counted terms, we still need to check that we won't try to access
characters beyond the first character of the subject string.  For functions like readSurrogatePairChecked(),
we check for input before calling the funcion.  For new input functions with a try prefix like tryReadBackward,
the function itselfs checks for available input.  After these checks prove that it is safe to access an offset
to the left of the current input position, the actual matching can be performed.

The Yarr parser, parses regular expression in left to right order.  It also computes character offest in forward
order.  When we Byteterm compile, we process backward matching disjunctions right to left.  The parser also has
special handling of forward references within a backward matching parenthetical group.  All such forward references
are saved for that parenthetical group and are processed at the end of the group.  Every one of these forward
reference are check to see if a capture to the right of the forward reference was found, if so the forward
reference is converted to a back reference.

As part of this work, the ByteTerm dumping code was significantly updated to allow for not only dumping of the
ByteCode after it has been generated, but to dump ByteCode while it is being interpreted.  This ByteTerm dumping
while interpreting is enabled with the Interpreter::verbose compile time constant.

Reviewed by Yusuke Suzuki.

* JSTests/stress/regexp-lookbehind.js: New tests.
(arrayToString):
(dumpValue):
(compareArray):
(testRegExp):
* JSTests/test262/config.yaml:
* Source/JavaScriptCore/runtime/RegExp.cpp:
(JSC::RegExp::compile):
(JSC::RegExp::compileMatchOnly):
* Source/JavaScriptCore/yarr/YarrInterpreter.cpp:
(JSC::Yarr::ByteTermDumper::ByteTermDumper):
(JSC::Yarr::ByteTermDumper::unicode):
(JSC::Yarr::Interpreter::InputStream::readForCharacterDump):
(JSC::Yarr::Interpreter::InputStream::tryReadBackward):
(JSC::Yarr::Interpreter::InputStream::tryUncheckInput):
(JSC::Yarr::Interpreter::InputStream::isValidNegativeInputOffset):
(JSC::Yarr::Interpreter::InputStream::dump const):
(JSC::Yarr::Interpreter::checkCharacter):
(JSC::Yarr::Interpreter::checkSurrogatePair):
(JSC::Yarr::Interpreter::checkCasedCharacter):
(JSC::Yarr::Interpreter::checkCharacterClass):
(JSC::Yarr::Interpreter::checkCharacterClassDontAdvanceInputForNonBMP):
(JSC::Yarr::Interpreter::tryConsumeBackReference):
(JSC::Yarr::Interpreter::matchAssertionWordBoundary):
(JSC::Yarr::Interpreter::backtrackPatternCharacter):
(JSC::Yarr::Interpreter::backtrackPatternCasedCharacter):
(JSC::Yarr::Interpreter::matchCharacterClass):
(JSC::Yarr::Interpreter::backtrackCharacterClass):
(JSC::Yarr::Interpreter::matchBackReference):
(JSC::Yarr::Interpreter::backtrackBackReference):
(JSC::Yarr::Interpreter::recordParenthesesMatch):
(JSC::Yarr::Interpreter::matchParenthesesOnceBegin):
(JSC::Yarr::Interpreter::matchParenthesesOnceEnd):
(JSC::Yarr::Interpreter::backtrackParenthesesOnceEnd):
(JSC::Yarr::Interpreter::matchParentheticalAssertionBegin):
(JSC::Yarr::Interpreter::backtrackParentheticalAssertionBegin):
(JSC::Yarr::Interpreter::matchDisjunction):
(JSC::Yarr::ByteCompiler::compile):
(JSC::Yarr::ByteCompiler::haveCheckedInput):
(JSC::Yarr::ByteCompiler::assertionWordBoundary):
(JSC::Yarr::ByteCompiler::atomPatternCharacter):
(JSC::Yarr::ByteCompiler::atomCharacterClass):
(JSC::Yarr::ByteCompiler::atomBackReference):
(JSC::Yarr::ByteCompiler::atomParenthesesOnceBegin):
(JSC::Yarr::ByteCompiler::atomParenthesesTerminalBegin):
(JSC::Yarr::ByteCompiler::atomParenthesesSubpatternBegin):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionBegin):
(JSC::Yarr::ByteCompiler::atomParentheticalAssertionEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesSubpatternEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesOnceEnd):
(JSC::Yarr::ByteCompiler::atomParenthesesTerminalEnd):
(JSC::Yarr::ByteCompiler::emitDisjunction):
(JSC::Yarr::ByteCompiler::isSafeToRecurse):
(JSC::Yarr::ByteTermDumper::dumpTerm):
(JSC::Yarr::ByteTermDumper::dumpDisjunction):
(JSC::Yarr::Interpreter::InputStream::readPair): Deleted.
(JSC::Yarr::ByteCompiler::dumpDisjunction): Deleted.
* Source/JavaScriptCore/yarr/YarrInterpreter.h:
(JSC::Yarr::ByteTerm::ByteTerm):
(JSC::Yarr::ByteTerm::HaveCheckedInput):
(JSC::Yarr::ByteTerm::WordBoundary):
(JSC::Yarr::ByteTerm::BackReference):
(JSC::Yarr::ByteTerm::isCharacterType):
(JSC::Yarr::ByteTerm::isCasedCharacterType):
(JSC::Yarr::ByteTerm::isCharacterClass):
(JSC::Yarr::ByteTerm::matchDirection):
* Source/JavaScriptCore/yarr/YarrJIT.cpp:
(JSC::Yarr::dumpCompileFailure):
* Source/JavaScriptCore/yarr/YarrJIT.h:
* Source/JavaScriptCore/yarr/YarrParser.h:
(JSC::Yarr::Parser::parseParenthesesBegin):
* Source/JavaScriptCore/yarr/YarrPattern.cpp:
(JSC::Yarr::YarrPatternConstructor::resetForReparsing):
(JSC::Yarr::YarrPatternConstructor::assertionBOL):
(JSC::Yarr::YarrPatternConstructor::atomPatternCharacter):
(JSC::Yarr::YarrPatternConstructor::atomBuiltInCharacterClass):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesSubpatternBegin):
(JSC::Yarr::YarrPatternConstructor::atomParentheticalAssertionBegin):
(JSC::Yarr::YarrPatternConstructor::atomParenthesesEnd):
(JSC::Yarr::YarrPatternConstructor::atomBackReference):
(JSC::Yarr::YarrPatternConstructor::copyDisjunction):
(JSC::Yarr::YarrPatternConstructor::quantifyAtom):
(JSC::Yarr::YarrPatternConstructor::disjunction):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::SavedContext::SavedContext):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::SavedContext::restore):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::ParenthesisContext):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::push):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::pop):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::setInvert):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::invert const):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::setMatchDirection):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::matchDirection const):
(JSC::Yarr::YarrPatternConstructor::ParenthesisContext::reset):
(JSC::Yarr::YarrPatternConstructor::pushParenthesisContext):
(JSC::Yarr::YarrPatternConstructor::popParenthesisContext):
(JSC::Yarr::YarrPatternConstructor::setParenthesisInvert):
(JSC::Yarr::YarrPatternConstructor::parenthesisInvert const):
(JSC::Yarr::YarrPatternConstructor::setParenthesisMatchDirection):
(JSC::Yarr::YarrPatternConstructor::parenthesisMatchDirection const):
(JSC::Yarr::YarrPattern::YarrPattern):
(JSC::Yarr::dumpCharacterClass):
(JSC::Yarr::PatternTerm::dump):
* Source/JavaScriptCore/yarr/YarrPattern.h:
(JSC::Yarr::PatternTerm::PatternTerm):
(JSC::Yarr::PatternTerm::convertToBackreference):
(JSC::Yarr::PatternTerm::setMatchDirection):
(JSC::Yarr::PatternTerm::matchDirection const):
(JSC::Yarr::PatternAlternative::PatternAlternative):
(JSC::Yarr::PatternAlternative::matchDirection const):
(JSC::Yarr::PatternDisjunction::addNewAlternative):
(JSC::Yarr::YarrPattern::resetForReparsing):
* Source/JavaScriptCore/yarr/YarrSyntaxChecker.cpp:
(JSC::Yarr::SyntaxChecker::atomParentheticalAssertionBegin):
* Source/WTF/wtf/PrintStream.cpp:
(WTF::printInternal):
* Source/WTF/wtf/PrintStream.h:
* Source/WebCore/contentextensions/URLFilterParser.cpp:
(WebCore::ContentExtensions::PatternParser::atomParentheticalAssertionBegin):

Canonical link: https://commits.webkit.org/257823@main
